File: src/move_it/simple_moveit_example.py
#!/usr/bin/env python

import sys
import time
import rospy
import moveit_commander
import moveit_msgs.msg
import geometry_msgs.msg
from math import pi
import math
from std_srvs.srv import Empty
import copy
from moveit_commander.conversions import pose_to_list, list_to_pose

class ExampleMoveItTrajectories(object):

    def __init__(self):

        #Initialize the node
        moveit_commander.roscpp_initialize(sys.argv)
        rospy.init_node('example_move_it_trajectories')

        #Create the MoveItInterface necessary objects
        arm_group_name="arm"
        self.robot = moveit_commander.RobotCommander()
        self.scene = moveit_commander.PlanningSceneInterface()

        #Create arm group module
        self.arm_group = moveit_commander.MoveGroupCommander(arm_group_name)
        display_trajectory_publisher = rospy.Publisher('move_group/display_planned_path',
                                                        moveit_msgs.msg.DisplayTrajectory,
                                                        queue_size=20)
        #Create gripper group module
        gripper_group_name="gripper"
        self.gripper_group = moveit_commander.MoveGroupCommander(gripper_group_name)
	gripper_joint_names = rospy.get_param(rospy.get_namespace() + "gripper_joint_names", [])
        self.gripper_joint_name = gripper_joint_names[0]




    
    def reach_named_position(self,target):
        # [home,vertical]
        arm_group=self.arm_group

        # Going to one of those targets
        rospy.loginfo("Going to named target " + target)
        # Set the target
        arm_group.set_named_target(target)
        # Plan the trajectory
        planned_path1 = arm_group.plan()
        # Execute the trajectory and block while it's not finished
        arm_group.execute(planned_path1, wait=True)
            
    def reach_joint_angles(self,tolerance):

        arm_group = self.arm_group
        success = True

        # Get the current joint positions
        joint_positions = arm_group.get_current_joint_values()
        rospy.loginfo("Printing current joint positions before movement :")

        # Set the goal joint tolerance
        self.arm_group.set_goal_joint_tolerance(tolerance)

        # Set the joint target configuration
        joint_positions[0] = 0
        joint_positions[1] = 0
        joint_positions[2] = pi/2
        joint_positions[3] = pi/4
        joint_positions[4] = 0
        joint_positions[5] = pi/2
        arm_group.set_joint_value_target(joint_positions)
        
        # Plan and execute in one command
        arm_group.go(wait=True)

    def get_cartesian_pose(self):
        arm_group = self.arm_group

        # Get the current pose
        pose = arm_group.get_current_pose()

        return pose.pose


    def reach_gripper_position(self, relative_position):
            gripper_group = self.gripper_group
            
            # We only have to move this joint because all others are mimic!
            gripper_joint = self.robot.get_joint(self.gripper_joint_name)
            gripper_max_absolute_pos = gripper_joint.max_bound()
            gripper_min_absolute_pos = gripper_joint.min_bound()
            try:
                val = gripper_joint.move(relative_position * (gripper_max_absolute_pos - gripper_min_absolute_pos) + gripper_min_absolute_pos, True)
                return val
            except:
                return False 

    def compute_cartesian_path(self,waypoints):
            arm_group=self.arm_group
            (plan,fraction)=arm_group.compute_cartesian_path(waypoints,0.01,0.0)
            arm_group.execute(plan,wait=True)
            rospy.sleep(3)

            
    def draw_circle(self):
            rospy.loginfo("Circle")
            waypoints = []

            # start with the current pose
            start_pose=self.get_cartesian_pose()
            start_pose.orientation.w=1.0

            for i in range(36):
              wpose=copy.deepcopy(start_pose)
              wpose.position.x=start_pose.position.x+0.1*math.cos(i*0.2)
              wpose.position.y=start_pose.position.y+0.1*math.sin(i*0.2)
              waypoints.append(copy.deepcopy(wpose))

            self.compute_cartesian_path(waypoints)
            self.reach_named_position("home")
      
    def draw_rectangle(self):

            rospy.loginfo("Rectangle")
            waypoints=[]

            #start with the current pose
            start_pose=self.get_cartesian_pose()
            start_pose.orientation.w=1.0

            wpose=copy.deepcopy(start_pose)
            wpose.position.x-=0.2
            waypoints.append(copy.deepcopy(wpose))

            wpose.position.y-=0.2
            waypoints.append(copy.deepcopy(wpose))

            wpose.position.x+=0.2
            waypoints.append(copy.deepcopy(wpose))

            wpose.position.y+=0.2
            waypoints.append(copy.deepcopy(wpose))

            self.compute_cartesian_path(waypoints)
            self.reach_named_position("home")

    def draw_triangle(self):
            #Triangle
            rospy.loginfo("Triangle")
            waypoints=[]

            #start with the current pose
            start_pose=self.get_cartesian_pose()
            start_pose.orientation.w=1.0

            wpose=copy.deepcopy(start_pose)
            wpose.position.x-=0.1
            wpose.position.y-=0.1*math.sqrt(3)
            waypoints.append(copy.deepcopy(wpose))

            wpose.position.x+=0.2
            waypoints.append(copy.deepcopy(wpose))

            wpose.position.x-=0.1
            wpose.position.y+=0.1*math.sqrt(3)
            waypoints.append(copy.deepcopy(wpose))

            self.compute_cartesian_path(waypoints)
            self.reach_named_position("home")

    def draw_a(self):
            #Draw A
            #start with the current pose
            rospy.loginfo("Draw Alphabet A")
            waypoints=[]
            center_pose=self.get_cartesian_pose()
            center_pose.orientation.w=1.0

            wpose=copy.deepcopy(center_pose)
            wpose.position.x-=0.1
            wpose.position.y-=0.1*math.sqrt(3)
            waypoints.append(copy.deepcopy(wpose))


            wpose.position.x+=0.1
            wpose.position.y+=0.1*math.sqrt(3)
            waypoints.append(copy.deepcopy(wpose))

            wpose.position.x+=0.1
            wpose.position.y-=0.1*math.sqrt(3)
            waypoints.append(copy.deepcopy(wpose))


            wpose.position.x-=0.1
            wpose.position.y+=0.1*math.sqrt(3)
            waypoints.append(copy.deepcopy(wpose))

            wpose.position.x-=0.05
            wpose.position.y-=0.05*math.sqrt(3)
            waypoints.append(copy.deepcopy(wpose))

            wpose.position.x+=0.1
            waypoints.append(copy.deepcopy(wpose))
            self.compute_cartesian_path(waypoints)
            self.reach_named_position("home")







def main():

  example=ExampleMoveItTrajectories()
  raw_input()
  example.reach_gripper_position(0.5)


  #For testing purposes
  example.reach_named_position("home")
  

  #Circle
  example.draw_circle()

  # Rectangle
  example.draw_rectangle()  

  # Triangle
  example.draw_triangle()

  # Draw A
  example.draw_a()  





if __name__=='__main__':
  main()

File: src/move_it/simple_moveit_example.cpp
/*********************************************************************
 * Software License Agreement (BSD License)
 *
 *  Copyright (c) 2013, SRI International
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials provided
 *     with the distribution.
 *   * Neither the name of SRI International nor the names of its
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 *********************************************************************/

/* Author: Sachin Chitta, Dave Coleman, Mike Lautman */

#include <moveit/move_group_interface/move_group_interface.h>
#include <moveit/planning_scene_interface/planning_scene_interface.h>

#include <moveit_msgs/DisplayRobotState.h>
#include <moveit_msgs/DisplayTrajectory.h>

#include <moveit_msgs/AttachedCollisionObject.h>
#include <moveit_msgs/CollisionObject.h>

#include <cmath>
using namespace std;
// The circle constant tau = 2*pi. One tau is one rotation in radians.
const double tau = 2 * M_PI;




void go_to_pose_goal(moveit::planning_interface::MoveGroupInterface &move_group_interface) {
  // .. _move_group_interface-planning-to-pose-goal:
  //
  // Planning to a Pose goal
  // ^^^^^^^^^^^^^^^^^^^^^^^
  // We can plan a motion for this group to a desired pose for the
  // motion planning by controlling end-effector. 
  geometry_msgs::Pose target_pose1;
  target_pose1.orientation.w = 1.0;
  target_pose1.position.x = 0.28;
  target_pose1.position.y = -0.2;
  target_pose1.position.z = 0.5;
  move_group_interface.setPoseTarget(target_pose1);


  moveit::planning_interface::MoveGroupInterface::Plan my_plan;

  bool success = (move_group_interface.plan(my_plan) == moveit::planning_interface::MoveItErrorCode::SUCCESS);

  // Finally, to execute the trajectory stored in my_plan, you could use the following method call:
  move_group_interface.execute(my_plan);

}


void  go_to_joint_state 
(
  moveit::planning_interface::MoveGroupInterface &move_group_interface,
  const moveit::core::JointModelGroup* joint_model_group,
  moveit::core::RobotStatePtr &initial_pose
)
// Planning to a joint state
  // ^^^^^^^^^^^^^^^^^^^^^^^
  // We can plan a motion for this group to a desired pose by controlling the angles of joints.
{

  moveit::planning_interface::MoveGroupInterface::Plan my_plan;

  // Next get the current set of joint values for the group.
  std::vector<double> joint_group_positions;
  initial_pose->copyJointGroupPositions(joint_model_group, joint_group_positions);
  // Now, let's modify one of the joints, plan to the new joint space goal and visualize the plan.
  move_group_interface.setJointValueTarget(joint_group_positions);

  bool success = (move_group_interface.plan(my_plan) == moveit::planning_interface::MoveItErrorCode::SUCCESS);
  move_group_interface.execute(my_plan);
}

// Cartesian Path Planning
// Cartesian Planning:  the generation of motion trajectories where a goal is specified 
// in terms of the desired location of the end effector
void draw_rectangle(
  moveit::planning_interface::MoveGroupInterface &move_group_interface) {

 // Cartesian Rectangle
 // We will draw rectangle on the xy-plane.
  moveit_msgs::RobotTrajectory trajectory;
 std::vector<geometry_msgs::Pose> waypoints_rectangle;
 geometry_msgs::Pose start=move_group_interface.getCurrentPose().pose;

start.position.y-=0.1;
 waypoints_rectangle.push_back(start);
 start.position.x+=0.1;
 waypoints_rectangle.push_back(start);
 start.position.y+=0.1;
 waypoints_rectangle.push_back(start);
 start.position.x-=0.1;
 waypoints_rectangle.push_back(start);
 
// with end points of end-effector in waypoints and computerCartesianPath function, you can plan Cartesian path.
double fraction = move_group_interface.computeCartesianPath(waypoints_rectangle, 0.01, 0.0, trajectory);
move_group_interface.execute(trajectory);
}


// Cartesian Path Planning
// Cartesian Planning:  the generation of motion trajectories where a goal is specified 
// in terms of the desired location of the end effector
void draw_circle(moveit::planning_interface::MoveGroupInterface &move_group_interface) {

  // We will draw that the center is current state and 0.1m radius on the xy-plane.
  //define center of circle
  double radius=0.1;
  double pi=3.14159;
  moveit_msgs::RobotTrajectory trajectory;
  std::vector<geometry_msgs::Pose> waypoints_circle;
  geometry_msgs::Pose center=move_group_interface.getCurrentPose().pose;

  geometry_msgs::Pose start=move_group_interface.getCurrentPose().pose;

  for (double angle=2*pi;angle>0;angle-=0.2) {
    start.position.x=center.position.x+radius*sin(angle);
    start.position.y=center.position.y+radius*cos(angle);
    waypoints_circle.push_back(start);
  }

  // with end points of end-effector in waypoints and computerCartesianPath function, you can plan Cartesian path.
  double fraction = move_group_interface.computeCartesianPath(waypoints_circle, 0.01, 0.0, trajectory);
  move_group_interface.execute(trajectory);
}


// Cartesian Path Planning
// Cartesian Planning:  the generation of motion trajectories where a goal is specified 
// in terms of the desired location of the end effector
void draw_A(moveit::planning_interface::MoveGroupInterface &move_group_interface) {

// We will draw A on the xy-plane.
moveit_msgs::RobotTrajectory trajectory;
std::vector<geometry_msgs::Pose> waypoints_A;
geometry_msgs::Pose start=move_group_interface.getCurrentPose().pose;

start.position.x-=0.1;
start.position.y-=0.1*sqrt(3);
waypoints_A.push_back(start);

start.position.x+=0.1;
start.position.y+=0.1*sqrt(3);
waypoints_A.push_back(start);

start.position.x+=0.1;
start.position.y-=0.1*sqrt(3);
waypoints_A.push_back(start);

start.position.x-=0.1;
start.position.y+=0.1*sqrt(3);
waypoints_A.push_back(start);

start.position.x-=0.05;
start.position.y-=0.05*sqrt(3);
waypoints_A.push_back(start);

start.position.x+=0.1;
waypoints_A.push_back(start);

  // with end points of end-effector in waypoints and computerCartesianPath function, you can plan Cartesian path.
  double fraction = move_group_interface.computeCartesianPath(waypoints_A, 0.01, 0.0, trajectory);
  move_group_interface.execute(trajectory);
}

void back_to_home(moveit::planning_interface::MoveGroupInterface &move_group_interface,
geometry_msgs::Pose &initial_pose) {

moveit_msgs::RobotTrajectory trajectory;
std::vector<geometry_msgs::Pose> waypoints;
geometry_msgs::Pose start=move_group_interface.getCurrentPose().pose;
waypoints.push_back(start);
waypoints.push_back(initial_pose);

// with end points of end-effector in waypoints and computerCartesianPath function, you can plan Cartesian path.
double fraction = move_group_interface.computeCartesianPath(waypoints, 0.01, 0.0, trajectory);
move_group_interface.execute(trajectory);

}
int main(int argc, char** argv)
{
  ros::init(argc, argv, "move_group_interface_tutorial");
  ros::NodeHandle node_handle;

  // ROS spinning must be running for the MoveGroupInterface to get information
  // about the robot's state. One way to do this is to start an AsyncSpinner
  // beforehand.
  ros::AsyncSpinner spinner(1);
  spinner.start();


  // MoveIt operates on sets of joints called "planning groups" and stores them in an object called
  // the `JointModelGroup`. Throughout MoveIt the terms "planning group" and "joint model group"
  // are used interchangeably.
  static const std::string PLANNING_GROUP = "arm";

 // The :planning_interface:`MoveGroupInterface` class can be easily
  // setup using just the name of the planning group you would like to control and plan for.
  moveit::planning_interface::MoveGroupInterface move_group_interface(PLANNING_GROUP);

  // We will use the :planning_interface:`PlanningSceneInterface`
  // class to add and remove collision objects in our "virtual world" scene
  moveit::planning_interface::PlanningSceneInterface planning_scene_interface;

  // Raw pointers are frequently used to refer to the planning group for improved performance.
  const moveit::core::JointModelGroup* joint_model_group =
      move_group_interface.getCurrentState()->getJointModelGroup(PLANNING_GROUP);

 // Initial joint values We will use this to come back to home position.
  moveit::core::RobotStatePtr initial_pose = move_group_interface.getCurrentState();

// Let's get simple moving

go_to_pose_goal(move_group_interface);
// Go to Home Position
go_to_joint_state(move_group_interface,joint_model_group,initial_pose);


cout<<"Draw Circle"<<endl;
draw_circle(move_group_interface);
// Go to Home Position
go_to_joint_state(move_group_interface,joint_model_group,initial_pose);

cout<<"Draw Rectangle"<<endl;
draw_rectangle(move_group_interface);
// Go to Home Position
go_to_joint_state(move_group_interface,joint_model_group,initial_pose);

cout<<"Draw A"<<endl;
draw_A(move_group_interface);
// Go to Home Position
go_to_joint_state(move_group_interface,joint_model_group,initial_pose);

ros::shutdown();


  return 0;
}






